### PR TITLE
EL-4236: Dashboard: option to prevent widgets from being moved

### PR DESCRIPTION
**Pull request by JonathanToddMF**
#### Checklist

* [x] The contents of this PR are mine to submit. No third party code or other material is being committed.
* [x] New third party dependencies (if any) are linked via `package.json`. Third party dependencies are licensed under one of the following: Apache, MIT, BSD, Mozilla Public License, Eclipse Public License, or Oracle Binary Code License.
* [x] The code in this PR does not contain export-restricted encryption algorithms.

#### Ticket / Issue
https://portal.digitalsafe.net/browse/EL-4236

#### Description of Proposed Changes
1. Updated documentation to show new input for ux-dashboard-widget
2. Added a new e2e test page to cover widget a widget having pinned set to true 
3. Resizing: Added checks to see whether there are widgets surrounding the widget being resized, if there is a widget with pinned set to true next to the action widget, resizing is prevented in that direction.
4. Movement: Prevent moving of action widget over widget that is pinned. Also prevented the placeholder from shifting widgets with pinned set to true.
5. If widgets have pinned set to true on the top row they are prevented from being shifted in functions like shiftWidgets and shiftWidgetsUp
7. Added similar functionality from resizing to keyboard resizing and movement. 
8. Updated getSurroundingWidgets function to get surrounding widgets based on all 8 directions.

#### Documentation CI URL

https://pages.github.houston.softwaregrp.net/sepg-docs-qa/UXAspects_CI_UXAspects_EL-4236-dashboard-option-prevent-widgets-moved

#### Downstream Build(s)

[ux-aspects-micro-focus](https://jenkins.swinfra.net/job/SEPG/job/caf/view/Developer/job/caf~ux-aspects-micro-focus~EL-4236-dashboard-option-prevent-widgets-moved~CI/)

### CODE REVIEW

File: src/components/dashboard/widget/dashboard-widget.component.ts [hunk 1]
```diff
@@ -80,6 +80,9 @@ export class DashboardWidgetComponent implements OnInit, AfterViewInit, OnDestro
     /** Defines a function that returns an aria label for the widget */
     @Input() widgetAriaLabel: (widgets: DashboardWidgetComponent) => string | string = this.getDefaultAriaLabel;
 
+    /** Defines whether or not this widget can be moved. */
+    @Input() canMove: boolean = true;
```
Comment (donna-marijne) [inline src/components/dashboard/widget/dashboard-widget.component.ts:84]: This should be coerced to boolean, see https://github.houston.softwaregrp.net/caf/ux-aspects-team/wiki/Tips-and-tricks#coerce-inputs.

**Review by donna-marijne — CHANGES_REQUESTED** [review]
Thanks, unfortunately looks like there are several issues here.

1. Widgets can reset to the wrong place after being dragged over an immovable widget (including stacking).
    https://plnkr.co/edit/tQhYF49H2uKbKHqy
    [Image from Gyazo](https://gyazo.com/6c9d0fb8b9d3075fd8d257bd39df0cab)

2. Widget can be stacked on an immovable widget via the keyboard.
    https://plnkr.co/edit/tQhYF49H2uKbKHqy
    [Image from Gyazo](https://gyazo.com/0b5338dcb58deddc2e92476fcd3f2302)

3. Widget can be stacked on an immovable widget via drag & drop.
    https://plnkr.co/edit/mRTJSAxcGPmAKiDX
    [Image from Gyazo](https://gyazo.com/952dc20758878a0932c7653f2a086369)

4. Widget with canMove=false can be moved by another widget.
    https://plnkr.co/edit/mRTJSAxcGPmAKiDX
    [Image from Gyazo](https://gyazo.com/1f7cd8238a250529bfcf106bd2afe976)

File: docs/app/pages/components/components-sections/dashboard/dashboard/dashboard.component.html [hunk 2]
```diff
@@ -215,6 +215,7 @@ <h3 class="widget-title">Alert</h3>
     <tr uxd-api-property name="minRowSpan" type="number" defaultValue="1">Defines the minimum number of rows this widget should occupy.</tr>
     <tr uxd-api-property name="resizable" type="boolean">Defines whether or not this widget can be resized.</tr>
     <tr uxd-api-property name="autoPositioning" type="boolean" defaultValue="true">Defines whether or not this widget will be automatically repositioned to fill any vacant space above it.</tr>
+    <tr uxd-api-property name="pinned" type="boolean" defaultValue="false">Defines whether or not this widget is pinned to the top row and whether it can be moved from its location. When set to <code>true</code> this widget will need a <code>row</code> and <code>col</code> specified.</tr>
```
Comment (donna-marijne) [inline docs/app/pages/components/components-sections/dashboard/dashboard/dashboard.component.html:218]: As usual, please break long lines. I would not expect `row` and `col` to have to be specified for this to work. `col` should implicitly be 0 if it is pinned.

File: e2e/pages/app/dashboard/pinned-layout/pinned-layout.testpage.component.html [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                  [class.dashboard-grabbing]="dashboard.isGrabbing"
+                  [options]="options"
+                  [(layout)]="layout">
+
+                  <ux-dashboard-widget
+                  id="subscription-widget"
+                  name="Service"
+                  [colSpan]="2"
+                  [resizable]="true">
+                  <div class="widget-content"
```
Comment (donna-marijne) [inline e2e/pages/app/dashboard/pinned-layout/pinned-layout.testpage.component.html:15]: Indentation issue.

File: e2e/pages/app/dashboard/pinned-layout/pinned-layout.testpage.component.html [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+              </ux-dashboard-widget>
+
+              <ux-dashboard-widget
+                  id="users-widget"
+                  name="Users"
+                  [resizable]="true"
+                  [row]="0"
+                  [col]="2"
+                  pinned>
```
Comment (donna-marijne) [inline e2e/pages/app/dashboard/pinned-layout/pinned-layout.testpage.component.html:37]:
```suggestion
                  pinned="true">
```

File: e2e/pages/app/dashboard/pinned-layout/pinned-layout.testpage.component.html [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                              uxDashboardGrabHandle
+                              #handle2="ux-dashboard-grab-handle">
+                          <ux-icon name="pan"></ux-icon>
+                      </button>
+                  </div>
+
+
+              </ux-dashboard-widget>
+
+              <ux-dashboard-widget id="alert-widget" name="Alert" [col]="undefined" [row]="undefined">
```
Comment (donna-marijne) [inline e2e/pages/app/dashboard/pinned-layout/pinned-layout.testpage.component.html:54]:
```suggestion
              <ux-dashboard-widget id="alert-widget" name="Alert">
```

File: src/components/dashboard/widget/dashboard-widget.component.ts [hunk 6]
```diff
@@ -80,6 +80,15 @@ export class DashboardWidgetComponent implements OnInit, AfterViewInit, OnDestro
     /** Defines a function that returns an aria label for the widget */
     @Input() widgetAriaLabel: (widgets: DashboardWidgetComponent) => string | string = this.getDefaultAriaLabel;
 
+    /** Defines whether or not this widget can be moved from the top row. */
+    @Input() set pinned(pinned: boolean) {
+        this._pinned = coerceBooleanProperty(pinned);
```
Comment (donna-marijne) [inline src/components/dashboard/widget/dashboard-widget.component.ts:85]: You missed the `ngAcceptInputType_` property.

File: src/components/dashboard/dashboard.service.ts [hunk 7]
```diff
@@ -395,9 +419,11 @@ export class DashboardService implements OnDestroy {
...
 
             case ActionDirection.BottomLeft:
-                dimensions.height += mouseY;
-                dimensions.x += mouseX;
-                dimensions.width -= mouseX;
+                if (surroundingWidgetPinned.filter(wgt => wgt.row === action.widget.row).length === 0 || dimensions.width <= initialWidgetWidth + movementBuffer || centerX < mousePosX) {
+                    dimensions.x += mouseX;
+                    dimensions.width -= mouseX;
+                }
+                    dimensions.height += mouseY;
```
Comment (donna-marijne) [inline src/components/dashboard/dashboard.service.ts:426]: Indentation issue.

File: src/components/dashboard/dashboard.service.ts [hunk 8]
```diff
@@ -830,6 +878,34 @@ export class DashboardService implements OnDestroy {
             .map(space => space.widget);
     }
 
+    /**
+     * Check if a position in the dashboard is vacant or not
+     */
+    getMovedWidgetAvailablePosition(column: number, row: number, columnSpan: number, rowSpan: number, ignoreWidget?: DashboardWidgetComponent): boolean {
+
+
```
Comment (donna-marijne) [inline src/components/dashboard/dashboard.service.ts:886]: Remove excess newlines.

**Review by donna-marijne — CHANGES_REQUESTED** [review]
Thanks, seems to work well except for keyboard usage which still has the same issue with overlapping:
https://plnkr.co/edit/evFq3QUVslpVZePE?preview
[Image from Gyazo](https://gyazo.com/afeea3cf7784aff14955c3346b912c08)

Test coverage should be added to ensure that this does not happen.

Take care of code formatting as well.

**Comment by donna-marijne** [comment]
@JonathanToddMF As previously discussed, we have reached the limit of time that I'm willing to spend on this, so if the above changes can't be completed before the end of the day, let me know.

**Comment by JonathanToddMF** [comment]
Hey @donna-marijne all one of the above changes are fixed. The only remaining issue is that of specifying the col and row for the pinned widget. I think I should be able to resolve this before EOD.

**Review by donna-marijne — COMMENTED** [review]
I've started a review, but noticed another issue.
https://plnkr.co/edit/2tWjk1ufPHwzWL60
[Image from Gyazo](https://gyazo.com/c0f33aaedc40f906b698302544bd309c)

Unfortunately it's time to shelve this feature. Please keep the branch around, if we get to the bottom of the current backlog then we can come back to this.

**Pull request closed without merging**